ui/src/treegrid.tsx

```diff
@@ -0,0 +1,247 @@
+          </tbody>
+        </table>
+      </div>
+    </>
+  );
+}
+
+function renderChildren(
+  props: TreeGridProps,
+  state: TreeGridState,
+  toggleExpanded: (id: TreeGridId) => void,
+  id: TreeGridId,
+  indent: number,
+  collapse: boolean
+): JSX.Element[] {
+  const results: JSX.Element[] = [];
+
+  props.data[id]?.children?.forEach((childId) => {
+    const child = props.data[childId];
+    if (!child) {
+      return;
+    }
+    const childState = state.get(childId);
+
+    results.push(
+      <tr
+        key={childId}
+        style={{
+          ...(collapse && { visibility: "collapse" }),
+        }}
+      >
+        {props.columns.map((col, i) => {
+          let value = child.data[col.key] || "";
+          let title = "";
+          if (!(value instanceof Object)) {
+            value = String(value);
+            title = value;
+          }
+
+          return (
+            <td
+              key={i}
+              style={{
+                ...(col.width && {
+                  maxWidth: col.width,
+                  width: col.width,
+                  minWidth: col.width,
+                }),
+              }}
+            >
+              <div
+                style={{
+                  textOverflow: "ellipsis",
+                  whiteSpace: "nowrap",
+                  overflow: "hidden",
+                  ...(i === 0 && { paddingLeft: `${indent}em` }),
+                }}
+              >
+                <Typography
+                  variant="body1"
+                  noWrap
+                  title={title}
+                  style={{
+                    display: "inline",
+                  }}
+                >
+                  {i === 0 && (
+                    <>
+                      {child.children?.length && (
+                        <IconButton
+                          size="small"
+                          title={childState?.errorMessage}
+                          onClick={() => {
+                            toggleExpanded(childId);
+                          }}
+                        >
+                          {itemIcon(childState)}
+                        </IconButton>
+                      )}
+                      {props.prefixElements?.(childId, child.data)}
+                    </>
+                  )}
+                  {value}
+                </Typography>
+              </div>
+            </td>
+          );
+        })}
+      </tr>
+    );
+
+    results.push(
+      ...renderChildren(
+        props,
+        state,
+        toggleExpanded,
+        childId,
+        indent + 1,
+        collapse || childState?.status !== Status.Expanded
+      )
+    );
+  });
+  return results;
+}
+
+enum Status {
+  Collapsed,
+  Expanded,
+  Loading,
+  Failed,
+}
+
+type TreeGridItemState = {
+  status: Status;
+  loaded?: boolean;
+  errorMessage?: string;
+};
+
+type TreeGridState = Map<TreeGridId, TreeGridItemState>;
+
+function itemIcon(state?: TreeGridItemState): JSX.Element {
```

Comment:
nit: is there a reason not to make this a normal react function component?

Reply:
Done. I had just thought of it as a helper but it was basically a component already.